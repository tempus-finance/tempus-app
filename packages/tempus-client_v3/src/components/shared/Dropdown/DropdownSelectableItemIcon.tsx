import { FC } from 'react';
import Icon, { IconType } from '../Icon';

export interface DropdownSelectableItemIconProps {
  icon: IconType;
  // eslint-disable-next-line react/no-unused-prop-types
  value?: any;
}

const DropdownSelectableItemIcon: FC<DropdownSelectableItemIconProps> = props => {
  const { icon } = props;

  return (
    <span className="tc__dropdownItem__icon-container">
      <Icon type={icon} size={12} />
    </span>
  );
};

export default DropdownSelectableItemIcon;
