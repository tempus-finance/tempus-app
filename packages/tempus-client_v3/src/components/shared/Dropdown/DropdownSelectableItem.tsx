import { Children, FC, isValidElement, ReactElement, useCallback, useState } from 'react';
import ButtonWrapper from '../ButtonWrapper';
import Typography from '../Typography';
import { DropdownSelectableItemIconProps } from './DropdownSelectableItemIcon';
import './DropdownItem.scss';

export interface DropdownSelectableItemProps {
  label: string;
  value?: string;
  selectedValue?: string;
  onClick?: (value: string) => void;
  children?: ReactElement<DropdownSelectableItemIconProps> | ReactElement<DropdownSelectableItemIconProps>[];
}

const DropdownSelectableItem: FC<DropdownSelectableItemProps> = props => {
  const { label, value, selectedValue, onClick, children } = props;
  const childrenValues = Children.map(children, child => (isValidElement(child) ? child.props.value : null))?.filter(
    val => val !== null,
  );
  const alreadySelectedIconIndex = childrenValues ? childrenValues.findIndex(val => val === selectedValue) : null;

  const [selectedIconIndex, setSelectedIconIndex] = useState(
    alreadySelectedIconIndex !== null && alreadySelectedIconIndex > -1 ? alreadySelectedIconIndex : 0,
  );

  const resolvedValue = childrenValues && selectedIconIndex !== null ? childrenValues[selectedIconIndex] : value;
  const selected = selectedValue === resolvedValue;

  const handleClick = useCallback(() => {
    let updatedSelectedIconIndex = selectedIconIndex;

    if (selected && selectedIconIndex !== null) {
      updatedSelectedIconIndex = (selectedIconIndex + 1) % Children.count(children);
      setSelectedIconIndex(updatedSelectedIconIndex);
    }

    const newValue =
      childrenValues && updatedSelectedIconIndex !== null && childrenValues[updatedSelectedIconIndex]
        ? childrenValues[updatedSelectedIconIndex]
        : value;
    onClick?.(newValue);
  }, [selectedIconIndex, selected, childrenValues, value, onClick, children]);

  return (
    <div className="tc__dropdownItem tc__dropdownItem__selectable">
      <ButtonWrapper onClick={handleClick}>
        <Typography variant="body-secondary" weight={selected ? 'bold' : 'regular'}>
          {label}
        </Typography>
      </ButtonWrapper>
      {Children.map(children, (child, index) => selectedIconIndex === index && child)}
    </div>
  );
};

export default DropdownSelectableItem;
