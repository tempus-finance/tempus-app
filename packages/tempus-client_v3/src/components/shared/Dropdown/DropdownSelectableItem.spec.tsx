import { fireEvent, render } from '@testing-library/react';
import DropdownSelectableItem, { DropdownSelectableItemProps } from './DropdownSelectableItem';
import DropdownSelectableItemIcon from './DropdownSelectableItemIcon';

const defaultProps: DropdownSelectableItemProps = {
  label: 'Item',
  value: 'item',
};

const mockOnClick = jest.fn<void, [string]>();

const defaultSubject = (props: DropdownSelectableItemProps) => render(<DropdownSelectableItem {...props} />);

const defaultSubjectWithIcons = (props: DropdownSelectableItemProps) =>
  render(
    <DropdownSelectableItem {...props}>
      <DropdownSelectableItemIcon icon="up-arrow-thin" value="up" />
      <DropdownSelectableItemIcon icon="down-arrow-thin" value="down" />
    </DropdownSelectableItem>,
  );

const defaultSubjectWithIconsNoValues = (props: DropdownSelectableItemProps) =>
  render(
    <DropdownSelectableItem {...props}>
      <DropdownSelectableItemIcon icon="up-arrow-thin" />
      <DropdownSelectableItemIcon icon="down-arrow-thin" />
    </DropdownSelectableItem>,
  );

describe('DropdownSelectableItem', () => {
  it('renders a selectable dropdown item', () => {
    const { getByRole } = defaultSubject(defaultProps);

    const actualButton = getByRole('button');

    expect(actualButton).not.toBeNull();
    expect(actualButton).toMatchSnapshot();
  });

  it('renders a selectable dropdown item with icons', () => {
    const { container, getByRole } = defaultSubjectWithIcons(defaultProps);

    const actualButton = getByRole('button');
    const upIcon = container.querySelector('.tc__icon');

    expect(actualButton).not.toBeNull();
    expect(actualButton).toMatchSnapshot();

    expect(upIcon).not.toBeNull();
    expect(upIcon).toMatchSnapshot();
  });

  it('changes the icon on click', () => {
    const { container, getByRole } = defaultSubjectWithIcons({
      ...defaultProps,
      selectedValue: 'up',
      onClick: mockOnClick,
    });

    const actualButton = getByRole('button');
    let upIcon = container.querySelector('.tc__icon-up-arrow-thin');
    let downIcon = container.querySelector('.tc__icon-down-arrow-thin');

    expect(actualButton).not.toBeNull();

    expect(upIcon).not.toBeNull();
    expect(upIcon).toMatchSnapshot();

    expect(downIcon).toBeNull();

    fireEvent.click(actualButton);

    upIcon = container.querySelector('.tc__icon-up-arrow-thin');
    downIcon = container.querySelector('.tc__icon-down-arrow-thin');

    expect(mockOnClick).toBeCalledTimes(1);
    expect(mockOnClick).toBeCalledWith('down');

    expect(upIcon).toBeNull();

    expect(downIcon).not.toBeNull();
    expect(downIcon).toMatchSnapshot();
  });

  it('does not change the icon on click if the item was not selected', () => {
    const { container, getByRole } = defaultSubjectWithIcons({
      ...defaultProps,
      selectedValue: 'something', // different then 'up' or 'down'
      onClick: mockOnClick,
    });

    const actualButton = getByRole('button');
    let upIcon = container.querySelector('.tc__icon-up-arrow-thin');
    let downIcon = container.querySelector('.tc__icon-down-arrow-thin');

    expect(actualButton).not.toBeNull();

    expect(upIcon).not.toBeNull();
    expect(upIcon).toMatchSnapshot();

    expect(downIcon).toBeNull();

    fireEvent.click(actualButton);

    upIcon = container.querySelector('.tc__icon-up-arrow-thin');
    downIcon = container.querySelector('.tc__icon-down-arrow-thin');

    expect(mockOnClick).toBeCalledTimes(1);
    expect(mockOnClick).toBeCalledWith('up');

    expect(upIcon).not.toBeNull();
    expect(upIcon).toMatchSnapshot();

    expect(downIcon).toBeNull();
  });

  it('uses its value when `onClick` if the selected child does not have a value', () => {
    const { getByRole } = defaultSubjectWithIconsNoValues({
      ...defaultProps,
      selectedValue: 'something',
      onClick: mockOnClick,
    });

    const actualButton = getByRole('button');

    fireEvent.click(actualButton);

    expect(mockOnClick).toBeCalledTimes(1);
    expect(mockOnClick).toBeCalledWith('item');
  });
});
