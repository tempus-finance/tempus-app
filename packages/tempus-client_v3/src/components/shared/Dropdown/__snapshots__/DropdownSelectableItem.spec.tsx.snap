// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownSelectableItem renders a selectable dropdown item 1`] = `
<button
  class="tc__btn "
  data-selected="false"
  type="button"
>
  <div
    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
  >
    Item
  </div>
</button>
`;

exports[`DropdownSelectableItem renders a selectable dropdown item with icon 1`] = `
<svg
  class="tc__icon tc__icon-up-arrow-thin"
  fill="none"
  height="12"
  viewBox="0 0 12 12"
  width="12"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    clip-rule="evenodd"
    d="M5.61591 0.167368C5.82804 -0.0557892 6.17196 -0.0557892 6.38409 0.167368L9.8409 3.80392C10.053 4.02707 10.053 4.38888 9.8409 4.61204C9.62878 4.8352 9.28485 4.8352 9.07272 4.61204L6.54319 1.95098L6.54319 12H5.45681L5.45681 1.95098L2.92728 4.61204C2.71515 4.8352 2.37122 4.8352 2.1591 4.61204C1.94697 4.38888 1.94697 4.02707 2.1591 3.80392L5.61591 0.167368Z"
    fill="#222222"
    fill-rule="evenodd"
  />
</svg>
`;
